pkg/controller/database/postgres/postgres.go

```diff
@@ -0,0 +1,238 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane-contrib/provider-in-cluster/apis/database/v1alpha1"
+	"github.com/crossplane-contrib/provider-in-cluster/pkg/client/database/postgres"
+)
+
+const (
+	errUnexpectedObject                  = "the managed resource is not an Postgres resource"
+	ResourceCredentialsSecretDatabaseKey = "database"
+	errDelete                            = "failed to delete the Postgres resource"
+	errDeploymentMsg                     = "failed to get postgres deployment"
+	errServiceMsg                        = "failed to get postgres service"
+	errPVCCreateMsg                      = "failed to create or update postgres PVC"
+	errDeployCreateMsg                   = "failed to create or update postgres deployment"
+	errSVCCreateMsg                      = "failed to create or update postgres service"
+	errGeneratePasswordMsg               = "failed to generate potential postgres password"
+)
+
+// SetupPostgres adds a controller that reconciles Postgres instances.
+func SetupPostgres(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.PostgresGroupKind)
+	postgresLogger := l.WithValues("controller", name)
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Postgres{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.PostgresGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), newClientFn: postgres.NewRoleClient, logger: postgresLogger}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithLogger(postgresLogger),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube        client.Client
+	newClientFn func(kube client.Client) postgres.Client
+	logger      logging.Logger
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cfg, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	cs, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &external{client: c.newClientFn(c.kube), kube: c.kube, logger: c.logger, cs: cs}, nil
+}
+
+type external struct {
+	client postgres.Client
+	kube   client.Client
+	cs     kubernetes.Interface
+	logger logging.Logger
+}
+
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
+	ps, ok := mgd.(*v1alpha1.Postgres)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	//initial defaults
+	initializeDefaults(ps)
+
+	// check deployment status
+	dpl := &appsv1.Deployment{}
+	err := e.kube.Get(ctx, types.NamespacedName{Name: ps.Name, Namespace: ps.Namespace}, dpl)
+	if err != nil {
+		e.logger.Debug(errDeploymentMsg, "err", err)
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(func(err error) bool {
```

Comment:
Feel free to use this :) https://github.com/crossplane/crossplane-runtime/blob/67c39282bcb41ebed590398611693930ac779bc6/pkg/resource/resource.go#L193

Reply:
Thanks for sharing! I completely forgot about this :sweat_smile: